identifier-forest/src/compile_time_forest/mod.rs

```diff
@@ -0,0 +1,398 @@
+/// to be able to answer any kind of pattern.
+pub type CTForestGreedy6<I> = CTForest<I,
+    profile4::OGPSAlways,
+    profile4::SPOGAlways,
+    profile4::GPSOAlways,
+    profile4::POGSAlways,
+    profile4::GSPOAlways,
+    profile4::OSGPAlways
+>;
+
+/// A `CTForest` (Compile Time Forest) is a forest composed of 6 trees determined at
+/// compile time.
+///
+/// 6 trees is chosen because it is the minimal number to be able to have an optimal
+/// tree for each quad query pattern.
+///
+/// The chosen implementation of [`MaybeTree4`](crate::tree::MaybeTree4) is always
+/// [`OnceTreeSet`](crate::tree::OnceTreeSet).
+// TODO: used a non OnceCell base implementation for always instanciated tree?
+pub struct CTForest<I, Order1, Order2, Order3, Order4, Order5, Order6>
+where
+I: Identifier,
+Order1: Tree4Profile,
+// Order1::ALWAYS_INSTANCIATED = true
+Order2: Tree4Profile,
+Order3: Tree4Profile,
+Order4: Tree4Profile,
+Order5: Tree4Profile,
+Order6: Tree4Profile
+{
+    tree1: OnceTreeSet<I, Order1::First, Order1::Second, Order1::Third, Order1::Fourth>,
+    tree2: OnceTreeSet<I, Order2::First, Order2::Second, Order2::Third, Order2::Fourth>,
+    tree3: OnceTreeSet<I, Order3::First, Order3::Second, Order3::Third, Order3::Fourth>,
+    tree4: OnceTreeSet<I, Order4::First, Order4::Second, Order4::Third, Order4::Fourth>,
+    tree5: OnceTreeSet<I, Order5::First, Order5::Second, Order5::Third, Order5::Fourth>,
+    tree6: OnceTreeSet<I, Order6::First, Order6::Second, Order6::Third, Order6::Fourth>,
+}
+
+impl<I, Order1, Order2, Order3, Order4, Order5, Order6> CTForest<I, Order1, Order2, Order3, Order4, Order5, Order6>
+where
+I: Identifier,
+Order1: Tree4Profile,
+// Order1::ALWAYS_INSTANCIATED = true
+Order2: Tree4Profile,
+Order3: Tree4Profile,
+Order4: Tree4Profile,
+Order5: Tree4Profile,
+Order6: Tree4Profile
+{
+
+    pub fn new() -> Self {
+        Self {
+            tree1: make_once_tree_set::<I, Order1>(),
+            tree2: make_once_tree_set::<I, Order2>(),
+            tree3: make_once_tree_set::<I, Order3>(),
+            tree4: make_once_tree_set::<I, Order4>(),
+            tree5: make_once_tree_set::<I, Order5>(),
+            tree6: make_once_tree_set::<I, Order6>(),
+        }
+    }
+
+
+    fn best_conformance(&self, can_build: bool, pattern_layout: &[Option<I>; 4]) -> (usize, usize) {
+
+        [
+            self.tree1.index_conformance(can_build, pattern_layout),
+            self.tree2.index_conformance(can_build, pattern_layout),
+            self.tree3.index_conformance(can_build, pattern_layout),
+            self.tree4.index_conformance(can_build, pattern_layout),
+            self.tree5.index_conformance(can_build, pattern_layout),
+            self.tree6.index_conformance(can_build, pattern_layout),
+        ]
+            .iter()
+            .enumerate()
+            .filter(|v| v.1.is_some())
+            .max_by_key(|v| v.1.unwrap())
+            .map(|t| (t.0 + 1, t.1.unwrap()))
```

Comment:
```suggestion
            .filter_map(|(i, opt)| opt.map(|score| (i+1, score)))
            .max_by_key(|(_, score)| *score)
```
a little more compact and (IMO) idiomatic